packages/editor/src/store/actions.js

```diff
@@ -730,10 +904,47 @@ export function unlockPostSaving( lockName ) {
  *
  * @return {Object} Action object
  */
-export function resetEditorBlocks( blocks, options = {} ) {
+export function* resetEditorBlocks( blocks, options = {} ) {
+	const lastBlockAttributesChange = yield select( 'core/block-editor', 'getLastBlockAttributesChange' );
+
+	// Sync to sources from block attributes updates.
+	if ( lastBlockAttributesChange ) {
+		const updatedSources = new Set;
+
+		const blockStack = [ ...blocks ];
+
+		for ( let i = 0; i < blockStack.length; i++ ) {
+			const block = blockStack[ i ];
+
+			// Account for changes occurring within nested block.
+			if ( block.innerBlocks.length ) {
+				blockStack.push( ...block.innerBlocks );
+			}
+
+			if ( ! lastBlockAttributesChange.hasOwnProperty( block.clientId ) ) {
+				continue;
+			}
+
+			const blockType = yield select( 'core/blocks', 'getBlockType', block.name );
+			const changedAttributes = lastBlockAttributesChange[ block.clientId ];
+
+			for ( const [ attributeName, schema ] of Object.entries( blockType.attributes ) ) {
+				if ( changedAttributes.hasOwnProperty( attributeName ) && sources[ schema.source ] && sources[ schema.source ].update ) {
+					yield* sources[ schema.source ].update( schema, changedAttributes[ attributeName ] );
+					updatedSources.add( sources[ schema.source ] );
+				}
+			}
+		}
+
+		// Dependencies are reset so that source dependencies subscription
+		// skips a reset which would otherwise occur by dependencies change.
+		// This assures that at most one reset occurs per block change.
+		yield* resetLastBlockSourceDependencies( Array.from( updatedSources ) );
+	}
+
 	return {
 		type: 'RESET_EDITOR_BLOCKS',
-		blocks,
+		blocks: yield* getBlocksWithSourcedAttributes( blocks ),
 		shouldCreateUndoLevel: options.__unstableShouldCreateUndoLevel !== false,
 	};
 }
```

Comment:
To be completely fair, I got the idea of most of the functions here but it's hard to follow though. a schema or something somewhere might be good for maintenance.

Reply:
> To be completely fair, I got the idea of most of the functions here but it's hard to follow though. a schema or something somewhere might be good for maintenance.

I'd also wondered if I should try to do a better job here of extracting behaviors into functions, maybe also moving some of them into `block-sources/index.js`. Some of this could be future refactors as well.